### PR TITLE
Added method for customizable "boost-from-average"

### PR DESCRIPTION
**Pull request by olofer**
It may be possible to benefit from "boost-from-average" for many more objectives with an appropriate implementation of how the initial score should be computed (it is in general not simply the average label).

### CODE REVIEW

**Comment by chivee** [comment]
@guolinke , does it be consistence with parallel setting? 

@olofer , the custom logic will always enable for the xentropy settings?

**Comment by guolinke** [comment]
@chivee thanks for info..

@olofer you should sync up the global init score like this: https://github.com/Microsoft/LightGBM/blob/master/src/boosting/gbdt.cpp#L403-L425
can you create another PR to fix it ?

**Comment by olofer** [comment]
The custom average computation is only invoked if (1) boost_from_average is set, and (2) if the objective allows it at all (BoostFromAverage returns true) and (3) it is implemented and returns true. At least that was my interpretation. It seems to work the way I want on my test test scripts (using the CLI). The method can be used to improve this "auto-init-score" setting for the other objectives also.

OK I will have to look through the parallel stuff... thanks.

**Comment by chivee** [comment]
@olofer , I got you, what I'm concerning is that does this implementation will always work better than the original initial way, if the answer is no, we should added an option for the customization initial ways. @guolinke , maybe we can added add 'init_average' option.

for the parallel stuff , you can simply copy the logic.

**Comment by olofer** [comment]
Some thoughts: I think a scalar option "init_score" or "init_bias" (double) that allows a fully manual assignment to start from is the most flexible and should be available for usage (I think xgboost has this option); but it is super-convenient also to have these "auto_init" option, which could be plain average or the objective-specific average, but if the "init_score" is provided that one should override all other init settings...
